### PR TITLE
Chapter19

### PR DESCRIPTION
**Pull request by nazihaltajar**

### CODE REVIEW

File: Chapter9/Chapter9/AppDelegate.swift [hunk 1]
```diff
@@ -35,4 +36,30 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func applicationWillTerminate(_ application: UIApplication) {
     }
+    lazy var persistentContainer: NSPersistentContainer = {
```
Comment (adinas23) [inline Chapter9/Chapter9/AppDelegate.swift]: Leave an empty line

File: Chapter9/Chapter9/AppDelegate.swift [hunk 2]
```diff
@@ -35,4 +36,30 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+
+        let container = NSPersistentContainer(name: "Chapter9")
```
Comment (adinas23) [inline Chapter9/Chapter9/AppDelegate.swift:41]: Use a suggestive name for this container, no ```Chapter9```

File: Chapter9/Chapter9/AppDelegate.swift [hunk 3]
```diff
@@ -35,4 +36,30 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        container.loadPersistentStores(completionHandler: { (_, error) in
+            if let error = error as NSError? {
+
```
Comment (adinas23) [inline Chapter9/Chapter9/AppDelegate.swift]: Remove this empty line

File: Chapter9/Chapter9/AppDelegate.swift [hunk 4]
```diff
@@ -35,4 +36,30 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+    // MARK: - Core Data Saving support
+
+    func saveContext () {
+        let context = persistentContainer.viewContext
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+
```
Comment (adinas23) [inline Chapter9/Chapter9/AppDelegate.swift]: Remove the extra line

File: Chapter9/Chapter9/Controller/MapViewController.swift [hunk 5]
```diff
@@ -12,7 +12,7 @@ import MapKit
 class MapViewController: UIViewController {
     @IBOutlet weak private var mapView: MKMapView!
 
-    var restaurant = Restaurant()
+    var restaurant: RestaurantMO!
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/MapViewController.swift]: Has this controller every time a restaurantMO object?

File: Chapter9/Chapter9/Controller/NewRestaurantViewController.swift [hunk 6]
```diff
@@ -47,21 +63,34 @@ class NewRestaurantViewController: UITableViewController, UINavigationController
...
 
     override func viewDidLoad() {
         super.viewDidLoad()
 
         navigationController?.setBackButtonTintColor(mycolor: .white)
         customizationNavigationBar()
     }
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(true)
+        if let appDelegate = (UIApplication.shared.delegate as? AppDelegate) {
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/NewRestaurantViewController.swift]: You can extract appDelegate only once

File: Chapter9/Chapter9/Controller/NewRestaurantViewController.swift [hunk 7]
```diff
@@ -47,21 +63,34 @@ class NewRestaurantViewController: UITableViewController, UINavigationController
...
+            let request: NSFetchRequest<RestaurantMO> = RestaurantMO.fetchRequest()
+            let context = appDelegate.persistentContainer.viewContext
+            do { restaurants = try context.fetch(request)
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/NewRestaurantViewController.swift]: ```restaurants = try context.fetch(request)``` need to be on a new line

File: Chapter9/Chapter9/Controller/RestaurantTableViewController.swift [hunk 8]
```diff
@@ -7,44 +7,66 @@
...
-    func configure(withModel: Restaurant)
+    func configure(withModel: RestaurantMO)
 }
 
 class RestaurantTableViewController: UIViewController {
     @IBOutlet private weak var tableView: UITableView!
     @IBAction private func unwindToHome(segue: UIStoryboardSegue) {
         dismiss(animated: true, completion: nil)
     }
+    @IBOutlet var emptyRestaurantView: UIView!
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/RestaurantTableViewController.swift]: First you declare IBOutlets, after IBActions

File: Chapter9/Chapter9/Controller/RestaurantTableViewController.swift [hunk 9]
```diff
@@ -67,18 +89,16 @@ class RestaurantTableViewController: UIViewController {
...
         let checkInAction = UIContextualAction(style: .normal, title: "checkIn") {(_, _, completionHandler) in
-            let cell = tableView.cellForRow(at: indexPath)
-            self.restaurants[indexPath.row].isVisited = true
-            let imageView = UIImageView(image: UIImage(named: self.heartImageName))
-            cell?.accessoryView = imageView
+            self.restaurantMO[indexPath.row].isVisited = true
+            try? self.restaurantMO[indexPath.row].managedObjectContext?.save()
 
             completionHandler(true)
+
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/RestaurantTableViewController.swift]: Remove this empty line

File: Chapter9/Chapter9/Controller/RestaurantTableViewController.swift [hunk 10]
```diff
@@ -88,21 +108,26 @@ extension RestaurantTableViewController: UITableViewDelegate {
...
+                context.delete(restaurantToDelete)
+                appDelegate.saveContext()
+            }
 
             completionHandler(true)
         }
 
         let shareAction = UIContextualAction(style: .normal, title: shareText) {(_, _, completionHandler) in
-            let defaultText = "Just checking in at " + self.restaurants[indexPath.row].name
+            let defaultText = "Just checking in at " + self.restaurantMO[indexPath.row].name!
```
Comment (adinas23) [inline Chapter9/Chapter9/Controller/RestaurantTableViewController.swift]: Do you need this force unwrapped?

Comment (nazihaltajar) [reply Chapter9/Chapter9/Controller/RestaurantTableViewController.swift]: yes I do .. apparently

**Pull request closed without merging**